### PR TITLE
feat(lint): Adds support for subs with return type

### PR DESCRIPTION
**Pull request by davinci26**
Extends falco to support subroutines with return values

Fixes https://github.com/ysugimoto/falco/issues/8

Signed-off-by: Sotiris Nanopoulos <sotirisnan@gmail.com>

### CODE REVIEW

**Comment by davinci26** [comment]
It might have some rough edges left and right, will compare it tomorrow with our VCL corpus which is fairly big so it can expose bugs

**Comment by davinci26** [comment]
darn linter, I will fix this tomorrow

**Comment by davinci26** [comment]
I tried this PR on a fairly large corpus of VCL and it went ok. @ysugimoto wanna take a look? Honestly not 100% that I got the abstraction/code patterns all right but at least it works

**Comment by ysugimoto** [comment]
@davinci26 Thank you for your PR! I'll take a look in a few days 😃

File: context/context.go
Comment (ysugimoto) [inline context/context.go:547]: please remove debug code

File: context/context.go
Comment (ysugimoto) [inline context/context.go:553]:
```suggestion
			return nil, fmt.Errorf(`Function "%s" is not defined`, name)
```

File: context/context.go
Comment (ysugimoto) [inline context/context.go:548]:
```suggestion
		return nil, fmt.Errorf(`Function "%s" is not defined`, name)
```

File: linter/linter.go
Comment (ysugimoto) [inline linter/linter.go:281]: Conversely, we need to lint the fastly builtin subroutine (e.g `vcl_recv`) should not have a return type so we should lint them.

```vcl
sub vcl_recv BOOL {  // should lint error
  ...
}
```

Or, it can raise errors in the parser phase.

**Review by ysugimoto — CHANGES_REQUESTED** [review]
Basically, implementations are OK 👍 
But in syntax and VCL spec perspective, we need to know about the function subroutine.

I investigated for that and I understand its spec as follows:

----
In the function, the return value only could specify a literal expression.

```vcl
sub get_int INTEGER {
  declare local var.tmp INTEGER;
  set var.tmp = 10;
  return var.tmp  // invalid, could not specify variable
}
```
fiddle: https://fiddle.fastlydemo.net/fiddle/194ea094

----
The literal value could not enclose in parenthesis.

```vcl
sub get_int INTEGER {
  return (100); // invalid, literal could not enclose in parenthesis
}
```

fiddle: https://fiddle.fastlydemo.net/fiddle/d7fdc4c0

----
Return value MUST be a single expression.

```vcl
sub get_str STRING {
  declare local var.tmp STRING;
  set var.tmp = "foo";
  return (var.tmp "bar"); // invalid, return value must be a single literal (even enclosed in parenthesis or not)
}
```

fiddle: https://fiddle.fastlydemo.net/fiddle/968811cf

So the linter should satisfy their spec correctly.
When you noticed additional spec or use-case, let me know 🙃

**Comment by davinci26** [comment]
I have been travelling the past few days, I will address your comments today/tomorrow

File: linter/linter.go
Comment (davinci26) [reply linter/linter.go:286]: fixed in https://github.com/ysugimoto/falco/pull/57/commits/da98ba23e804ef06a32ef3367f3c35bc6b3f15e2

File: context/context.go
Comment (davinci26) [reply context/context.go]: lol 🤦 thanks

File: context/context.go
Comment (davinci26) [reply context/context.go]: done

**Comment by davinci26** [comment]
Regarding the big comment:

```
sub get_int INTEGER {
  declare local var.tmp INTEGER;
  set var.tmp = 10;
  return var.tmp  // invalid, could not specify variable
}
```

This is missing a semicolon but otherwise it should compile, https://fiddle.fastlydemo.net/fiddle/11869fd6

---
```
sub get_int INTEGER {
  return (100); // invalid, literal could not enclose in parenthesis
}
```

Already covered in https://github.com/ysugimoto/falco/blob/be9da0c50e8a8d6fa9ae029559abffeb5465a018/linter/linter_test.go#L1287

---
```
sub get_str STRING {
  declare local var.tmp STRING;
  set var.tmp = "foo";
  return (var.tmp "bar"); // invalid, return value must be a single literal (even enclosed in parenthesis or not)
}
```

This was a bit tricky to get it right, I implemented it in https://github.com/ysugimoto/falco/pull/57/commits/a80d30a36d6e618a3ac8bfa1752bca6b0b7b6254

---

**Review by ysugimoto — APPROVED** [review]
Sorry for being a late review, it looks fine!

Additionally, I put a comment about error rule matching.
After fixed it, this PR is ready to merge, thanks 😄

**Comment by davinci26** [comment]
> Sorry for being a late review, it looks fine!

No prob, thanks for the review :) Happy with the final version of the code

I think I have addressed your last comment, let me know if there is anything I can do ( I dont have merge perms so you will have to merge for me)

**Comment by ysugimoto** [comment]
@davinci26 Thanks, your last changes also look nice 👍 

I'll organize the latest release with other improvements! 
it takes a few days so please keep this PR open, I'll merge it when it's ready to release.

I'd like to add you to this project contributor. If you're interested in this project continuously, please accept our invitation 😄

**Comment by davinci26** [comment]
Yeah for sure :) Happy to help with reviews and the project!

**Pull request merged**
